cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/UserServiceEndpointStrategy.java

```diff
@@ -39,48 +39,71 @@
   private final ProgramId serviceId;
   private final String version;
   private final Random random;
+  private final RouteFallbackStrategy fallbackStrategy;
 
   public UserServiceEndpointStrategy(ServiceDiscovered serviceDiscovered, RouteStore routeStore, ProgramId serviceId,
-                                     @Nullable String version) {
+                                     String fallbackStrategy, @Nullable String version) {
     super(serviceDiscovered);
     this.routeStore = routeStore;
     this.serviceId = serviceId;
     this.version = version;
     this.random = ThreadLocalRandom.current();
+    this.fallbackStrategy = RouteFallbackStrategy.valueOfRouteFallbackStrategy(fallbackStrategy);
   }
 
-  public UserServiceEndpointStrategy(ServiceDiscovered serviceDiscovered, RouteStore routeStore, ProgramId serviceId) {
-    this(serviceDiscovered, routeStore, serviceId, null);
+  @VisibleForTesting
+  UserServiceEndpointStrategy(ServiceDiscovered serviceDiscovered, RouteStore routeStore, ProgramId serviceId) {
+    this(serviceDiscovered, routeStore, serviceId, null, null);
   }
 
   @Nullable
   @Override
   public Discoverable pick() {
+    String minMaxVersion = null;
     // If version is not null, then don't fetch any routeConfig.
     RouteConfig routeConfig = version == null ? routeStore.fetch(serviceId) : null;
+    boolean routeConfigUnAvailable = (routeConfig == null) || routeConfig.getRoutes().isEmpty();
```

Comment:
Usually a positive boolean variable is more preferred. E.g. in this case, better have `hasRouteConfig` instead. This helps the logic easier to read. Consider `if (hasRouteConfig)` vs `if (!routeConfigUnAvailable)`. Double negative is harder to understand.